### ISSUE TITLE
Stop using user-assignable type generics in the types of functions that don't do anything with them

### ISSUE THREAD

**Issue by Avaq** [fluture-js/Fluture#337]
This is an alternative to #230 made possible by the [introduction of the `unknown` type](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#new-unknown-top-type).

See also https://github.com/fluture-js/Fluture/issues/334#issuecomment-490139148.
See also https://github.com/fluture-js/Fluture/issues/370#issuecomment-533226653.

**Comment by Avaq** [comment]
TypeScript 3.4 introduces [higher order type inference from generic functions](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#higher-order-type-inference-from-generic-functions). This might make the need for using anything but an undetermined type variable/generic unnecessary. I should do some experiments with this.

**Comment by Avaq** [comment]
<sup>In response to https://github.com/fluture-js/Fluture/issues/370#issuecomment-533715945</sup>

I tried a simple thing myself, and it seems `never` and `unknown` have the same issues, but in opposite orders.

### With `never`/`undefined`

![latest-screenshot](https://user-images.githubusercontent.com/1217745/65394927-d1205980-dd93-11e9-86a7-5678ddabfbc1.png)
![latest-screenshot](https://user-images.githubusercontent.com/1217745/65394952-3aa06800-dd94-11e9-9e9c-d8f5a2647522.png)

Here the problem is that we're not allowed to __assign anything to `never`__ (the same goes for `undefined`). So when we're trying to assign `Future<string, never>` to `Future<never, number>`, it breaks on the *rejection branch*.

### With `unknown`

![latest-screenshot](https://user-images.githubusercontent.com/1217745/65394921-b9e16c00-dd93-11e9-95a6-83ac3118fead.png)

Here the problem is the other way around: we're not allowed to __assign `unknown` to anything__. So when we're trying to assign `Future<string, unknown>` to `Future<unknown, number>`, it breaks on the *resolution branch*.

----

In conclusion, it seems that neither `never` nor `unknown` possess the desired behaviour.

**Comment by Avaq** [comment]
It seems that `never` has the behaviour we want, but only when `chain` uses a union to define the type of the left branch. The downside is that this will allow users to mix their types through usage of `chain`.

**Comment by edgarjrg** [comment]
Yes i agree that neither of the types work right now.
